### DISCUSSION TITLE
Limitations of fast refresh?

### DISCUSSION THREAD

**Discussion by chrisdrackett** [vercel/next.js#12896] [Help] [answered]
Are there things that cause fast refresh to not work? for example is there a timer involved? I'm trying to debug an issue where we're not getting fast refresh, but as far as I can tell the only difference between our test (https://github.com/chrisdrackett/next-external-test) and actual project is that the actual project has many many more components.

**Comment by taylor-lindores-reeves** [comment]
Hey Chris, I have a big project and it's on 9.4.0, every time I save after running next dev it reloads the page after about 4 or 5 seconds. The little timer pops up in the bottom right. Basically, what I am saying, is that fast refresh is non-existent. Any idea why, or how it is switched on?

**Comment by jamesmosier** [comment]
Hi @chrisdrackett, is there a specific component that isn't refreshing like you'd expect? Fast Refresh worked on your repo as expected. I changed text and styles and the changes happened immediately without a full page refresh. 

Maybe try removing your `node_modules` and reinstall them? Maybe something went awry during the upgrade to 9.4? 🤔

**Comment by chrisdrackett** [comment]
yeah, my test repo is totally fine, but if I expand it out nothing in `common` works with fast refresh. I'm curious where to even begin to figure out what the issue is. I didn't want to spend a lot of time on this if there is a known limitation

**Reply by jamesmosier** [reply]
I'm not certain I understand "if I expand it out". What do you mean by that? 

I changed components and styling in `external` and `common` and saw the page update via Fast Refresh.

**Reply by chrisdrackett** [reply]
We just have a lot more components than in the example..

**Comment by Timer** [comment, marked as answer]
Fast Refresh will perform a full reload when you edit a file that's imported by modules outside of the React tree.

You might have a file which renders a React component but also exports a value that is imported by a non-React component. Consider migrating the non-React component export to a separate file and importing it into both files.

It is also possible you are using class components at the top-level of your application, which disables Fast Refresh. Fast Refresh requires at least one function component in your React tree.

**Reply by JClackett** [reply]
@Timer is it always going to be that nothing else can be exported from a file? as I think there are definitely use cases for it. My example is co-locating graphql queries + fragments inside the page that uses it. Moving the queries/fragments + any other things that belong in that file is quite a big no-no in my opinion as its just needless extra files :( what are your thoughts?

**Reply by JClackett** [reply]
@timneutkens any thoughts on the above?

**Reply by Timer** [reply]
Yes, this will always be the case while we are on webpack 4 or 5. This restriction is necessary due to webpack limitations.

**Comment by meoyawn** [comment]
same here. I have a small project with custom `_app.tsx`, `_document.tsx`, `tailwindcss` and have no idea why the page refreshes even on the smallest change. WIll try to scrap a small repro project

**Comment by meoyawn** [comment]
Hi @chrisdrackett @jamesmosier @Timer 

I've scraped together a small project that demonstrates full page refresh on every change to `src/pages/index.tsx` https://github.com/adelnizamutdinov/next-problem

Would greatly appreciate any comments and possible solutions to this

**Reply by timneutkens** [reply]
You're using an anonymous export here: https://github.com/adelnizamutdinov/next-problem/blob/master/src/pages/index.tsx#L4

If you change it to `export default function HomePage` instead it'll work

**Reply by meoyawn** [reply]
it works, thank you so much, you saved me tons of time! I shouldn't ignore `eslint-plugin-react`...

**Comment by pikeas** [comment]
Just bit by this as well. For anyone else who finds this issue while trying to figure out why Fast Reload isn't working, in addition to exporting a default function, the name of the function must be *capitalized* to qualify as a React component, otherwise you'll get a slow full refresh.

✅ `export default function Index() {...}`
❌ `export default function index() {...}`

Fast Refresh is great, please add some additional notes to https://nextjs.org/docs/basic-features/fast-refresh describing which patterns will disable it.

**Reply by Gyllsdorff** [reply]
> the name of the function must be capitalized to qualify as a React component


I ran into this one. I have a component that is wrapped by a [`connect()`](https://react-redux.js.org/api/connect) function from redux, since the root component is never used directly, except for tests, I named the component with a `_` prefix and the redux connected component without prefix so it would be easy to grep/search for and clear for the coder which was the exported component. 

The `_` prefix on the root component means NextJS thinks it something it can't do a fast refresh on so it reloads the page instead even if the `_` prefixed component isn't exported.

Smallest possible example, if you rename `_Foo` to `Foo1` then fast refresh works as expected.
```jsx
const _Foo = () => <p>test 2</p>
const Foo = _Foo
export default Foo
```

<details>
<summary>Triggers a full page reload</summary>
<p>

```jsx
const _Foo = ({ somethingFromRedux }) => <p>{somethingFromRedux} test 1</p>
const Foo = connect(s => ({ somethingFromRedux: s.something.from }))(_Foo)

// export { _Foo as Component }
export default Foo
```
<p>
</details>


<details>
<summary>In place, fast refresh</summary>
<p>

```jsx
const Foo = ({ somethingFromRedux }) => <p>{somethingFromRedux} test 1</p>
const ConnectedFoo = connect(s => ({ somethingFromRedux: s.something.from }))(Foo)

// export { Foo }
export default ConnectedFoo
```
<p>
</details>

**Reply by pingrishabh** [reply]
> Just bit by this as well. For anyone else who finds this issue while trying to figure out why Fast Reload isn't working, in addition to exporting a default function, the name of the function must be _capitalized_ to qualify as a React component, otherwise you'll get a slow full refresh.
> 
> ✅ `export default function Index() {...}` ❌ `export default function index() {...}`
> 
> Fast Refresh is great, please add some additional notes to https://nextjs.org/docs/basic-features/fast-refresh describing which patterns will disable it.

You just saved me hours. Kudos! 🔥

**Comment by jarrodpayne** [comment]
I'm trying to get Fast Refresh to work when a `node_modules` package is updated. Is this feasible? The full page is reloading when I update a dependency, but none of the rules above are being broken. For those wondering why I'd want to do such a thing: I'm working in a monorepo where my component library is a dependency for the next.js app.

**Reply by meoyawn** [reply]
check out https://github.com/martpie/next-transpile-modules, works for me

example: https://github.com/martpie/monorepo-typescript-next-the-sane-way

**Reply by jarrodpayne** [reply]
@meoyawn Thanks for send this my way. I ran this project, but changes to `shared/utils/calc.ts` trigger a full page reload instead of a Fast Refresh. I'm aware that transpiling `node_modules` can be achieved via `next-transpile-modules`, but I'm not looking for `node_modules` transpilation. I'm looking for Fast Refresh when a node_modules change is detected.

**Reply by jarrodpayne** [reply]
@meoyawn Apparently, `next-transpile-modules` is doing some magic to get Fast Refresh working for transpiled modules. I'll take a look at their code to see if I can figure out how. Thanks again.

**Comment by ariefurtado** [comment]
`next-transpile-modules` or builtin (after nextjs 13.1) config `transpilePackages: ['@my-monorepo/my-web-components']` do not worked for me. I am using stenciljs for my web-components library and the nrwl/nx to manage my monorepo.

For my specific case, I have a snippet that helps me to force a full page reload to ensure everything reloads. Its is definetly not the best of the worlds but it gets the job done for now...

the GIST: https://gist.github.com/ariefurtado/e2649317d5e770e9e882de040dc769d9
